include/bout/index_derivs.hxx

```diff
@@ -0,0 +1,789 @@
+ * This file is part of BOUT++.
+ *
+ * BOUT++ is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * BOUT++ is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with BOUT++.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ **************************************************************************/
+
+#ifndef __INDEX_DERIVS_HXX__
+#define __INDEX_DERIVS_HXX__
+
+#include <functional>
+#include <iostream>
+
+#include <bout/assert.hxx>
+#include <bout/constants.hxx>
+#include <bout/deriv_store.hxx>
+#include <bout/region.hxx>
+#include <bout/scorepwrapper.hxx>
+#include <bout/template_combinations.hxx>
+
+#include <bout_types.hxx>
+#include <fft.hxx>
+#include <msg_stack.hxx>
+#include <stencils.hxx>
+#include <unused.hxx>
+
+class Field3D;
+class Field2D;
+
+const BoutReal WENO_SMALL = 1.0e-8; // Small number for WENO schemes
+
+struct metaData {
+  // Would rather use a std::string here but this causes the
+  // metaData struct to be non-trivially destrucible which
+  // can prevent using temporary instances of this. Instead
+  // we'll use char* for now.
+  // const std::string key;
+  const char* key;
+  const int nGuards;
+  const DERIV derivType; // Can be used to identify the type of the derivative
+};
+
+/// Provide an easy way to report a Region's statistics
+inline std::ostream& operator<<(std::ostream& out, const metaData& meta) {
+  out << "key : " << meta.key;
+  out << ", ";
+  out << "nGuards : " << meta.nGuards;
+  out << ", ";
+  out << "type : " << DERIV_STRING(meta.derivType);
+  return out;
+}
+
+/// Here we define a helper class that provides a means to use a supplied
+/// stencil using functor to calculate a derivative over the entire field.
+/// Note we currently have a different interface for some of the derivative types
+/// to avoid needing different classes to represent the different operations
+/// The use of a functor here makes it possible to wrap up metaData into the
+/// type as well.
+template <typename FF>
+class DerivativeType {
+public:
+  template <DIRECTION direction, STAGGER stagger, int nGuards, typename T>
+  void standard(const T& var, T& result, REGION region) const {
+    TRACE("%s", __thefunc__);
+    ASSERT2(meta.derivType == DERIV::Standard || meta.derivType == DERIV::StandardSecond
+            || meta.derivType == DERIV::StandardFourth)
+    ASSERT2(var.getMesh()->template getNguard<direction>() >= nGuards);
+
+    BOUT_FOR(i, var.getRegion(region)) {
+      result[i] = apply(populateStencil<direction, stagger, nGuards>(var, i));
```

Comment:
This is awesome! So glad it's got to the point where the whole derivatives mess (as it was) boils down to something this "simple".